azure/services/privatedns/privatedns.go

```diff
@@ -20,202 +20,250 @@ import (
+
+		if !isVnetLinkManaged {
+			log.V(2).Info("Skipping vnet link deletion for unmanaged vnet link", "vnet link",
+				linkSpec.ResourceName(), "private dns zone", linkSpec.OwnerResourceName())
+			continue
 		}
-		if !isManaged {
-			log.V(1).Info("Skipping private DNS zone deletion for unmanaged private DNS zone", "private DNS", zoneSpec.ZoneName)
-			return nil
+
+		if err := s.vnetLinkReconciler.DeleteResource(ctx, linkSpec, serviceName); err != nil {
+			if !azure.IsOperationNotDoneError(err) || resErr == nil {
+				resErr = err
+			}
 		}
-		// Delete the private DNS zone, which also deletes all records.
-		log.V(2).Info("deleting private dns zone", "private dns zone", zoneSpec.ZoneName)
-		err = s.client.DeleteZone(ctx, s.Scope.ResourceGroup(), zoneSpec.ZoneName)
-		if err != nil && azure.ResourceNotFound(err) {
-			// already deleted
+	}
+
+	return resErr
+}
+
+func (s *Service) deleteZone(ctx context.Context, zoneSpec azure.ResourceSpecGetter) error {
+	ctx, log, done := tele.StartSpanWithLogger(ctx, "privatedns.Service.deleteZone")
+	defer done()
+
+	// Skip the reconciliation of private DNS zone which is not managed by capz.
+	isManaged, err := s.isPrivateDNSManaged(ctx, zoneSpec)
+	if err != nil {
+		if azure.ResourceNotFound(err) {
+			// already deleted or doesn't exist, cleanup status and return.
+			s.Scope.DeleteLongRunningOperationState(zoneSpec.ResourceName(), serviceName)
 			return nil
 		}
-		if err != nil && !azure.ResourceNotFound(err) {
-			return errors.Wrapf(err, "failed to delete private dns zone %s in resource group %s", zoneSpec.ZoneName, s.Scope.ResourceGroup())
-		}
-		log.V(2).Info("successfully deleted private dns zone", "private dns zone", zoneSpec.ZoneName)
+		return errors.Wrapf(err, "could not get private DNS zone state of %s in resource group %s", zoneSpec.ResourceName(), zoneSpec.ResourceGroupName())
 	}
-	return nil
+
+	if !isManaged {
+		log.V(1).Info("Skipping deletion of unmanaged private DNS zone", "private DNS", zoneSpec.ResourceName())
+		return nil
+	}
+
+	// Delete the private DNS zone, which also deletes all records
+	err = s.zoneReconciler.DeleteResource(ctx, zoneSpec, serviceName)
+	return err
 }
 
 // isPrivateDNSManaged returns true if the private DNS has an owned tag with the cluster name as value,
 // meaning that the DNS lifecycle is managed.
-func (s *Service) isPrivateDNSManaged(ctx context.Context, resourceGroup, zoneName string) (bool, error) {
-	zone, err := s.client.GetZone(ctx, resourceGroup, zoneName)
+func (s *Service) isPrivateDNSManaged(ctx context.Context, spec azure.ResourceSpecGetter) (bool, error) {
+	result, err := s.zoneGetter.Get(ctx, spec)
 	if err != nil {
 		return false, err
 	}
+	zone, ok := result.(privatedns.PrivateZone)
+	if !ok {
+		return false, errors.Errorf("%T is not a privatedns.PrivateZone", zone)
+	}
+
 	tags := converters.MapToTags(zone.Tags)
```

Comment:
You could probably refactor these last two lines to use your `isManaged` helper function since it just checks a map for `tags.HasOwned(clusterName)`.